sdk/azcore/etag.go

```diff
@@ -0,0 +1,62 @@
+// +build go1.13
+
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package azcore
+
+import (
+	"strings"
+)
+
+// ETag is a property used for optimistic concurrency during updates
+// ETag is a validator based on https://tools.ietf.org/html/rfc7232#section-2.3.2
+// An ETag can be empty ("") or nil. A nil etag represents the absence of an ETag, while an empty ETag is a valid ETag (e.g. https://www.rfc-editor.org/rfc/rfc7232.html#section-2.3)
+type ETag *string
+
+// StrongEquals does a strong comparison of two ETags. StrongEquals returns true when both
+// ETags are not weak and the values of the underlying strings are equal. If both ETags are "nil" they are considered equal
+func StrongEquals(a, b ETag) bool {
+	if !HasValue(a) || !HasValue(b) {
+		return a == b
+	}
+	return !IsWeak(a) && !IsWeak(b) && *a == *b
+}
+
+// WeakEquals does a weak compariosn of two ETags. Two ETags are equivalent if their opaque-tags match
+// character-by-character, regardless of either or both being tagged as "weak". If both ETags are "nil" they are considered equal
+func WeakEquals(a, b ETag) bool {
+	if !HasValue(a) || !HasValue(b) {
+		return a == b
+	}
+
+	getStart := func(e ETag) int {
+		if IsWeak(e) {
+			return 2
+		}
+		return 0
+	}
+	aStart := getStart(a)
+	bStart := getStart(b)
+
+	aVal := (*a)[aStart:]
+	bVal := (*b)[bStart:]
+
+	return aVal == bVal
+}
+
+// HasValue returns whether an ETag is present
+func HasValue(e ETag) bool {
```

Comment:
I feel like `HasValue()` and `IsWeak()` should be methods on `ETag` (maybe `WeakEquals()` to but less sure on that one).  Was this discussed?

Reply:
Oh yes, you're right. These should be methods. I don't know how I missed this.

Reply:
Go doesn't allow methods on Types that are pointers, we need to have `type ETag struct` instead of `type ETag *string` to do that
https://golang.org/ref/spec#Method_declarations

Reply:
I'd but the struct back then. The methods are better then functions.

Reply:
Seems like extra complexity, I'd favor `type ETag string`

Reply:
Let's say a method takes an ETag value for the if-match header. How does the method know whether to add the header with the etag value or not add the header at all if ETag is a string?